src/rust/engine/fs/src/store.rs

```diff
@@ -1738,6 +1738,47 @@ mod remote {
       })
     }
 
+    fn with_byte_stream_client<
+      Value: Send + 'static,
+      F: Fn(bazel_protos::bytestream_grpc::ByteStreamClient) -> BoxFuture<Value, String>
+        + Send
+        + Sync
+        + Clone
+        + 'static,
+    >(
+      &self,
+      f: F,
+    ) -> BoxFuture<Value, String> {
+      let store = self.clone();
+
+      future::loop_fn(0_usize, move |i| {
+        let ss = store.serverset.clone();
+        let f = f.clone();
+        store
+          .byte_stream_client()
+          .and_then(move |(client, callback_token)| {
+            future::ok(client)
+              .and_then(f)
+              .then(move |result| {
+                let health = match &result {
+                  &Ok(_) => serverset::Health::Healthy,
+                  &Err(_) => serverset::Health::Unhealthy,
+                };
+                ss.callback(callback_token, health);
+                result
+              }).to_boxed()
+          }).to_boxed()
+          .map(futures::future::Loop::Break)
+          .or_else(move |_| {
+            if i > 5 {
+              futures::future::err(format!("Too many loops {}", i))
```

Comment:
The error should probably explain why this would happen (not enough healthy servers?)

I feel like an argument could be made for putting the loop count and "giving up" into the serverset itself... because if you have no servers that qualify as healthy, does that really mean that you shouldn't try any of them? Or just increase your threshold?